### PR TITLE
427: configuration updates message broker

### LINKED ISSUE #427: Configuration propagation
Store configuration settings which can modify Storefront should be propagated to Storefront

AC:
* Create a new event which will send a message with Configuration data to Message Broker from Export API
  * collected configuration paths should be described in di.xml
  * only send a message if the value was changed
  * the initial command should collect and send all described configuration data
* Consumers(s) on the Message Broker side save the configuration to the corresponding Service via gRPC
  * configuration in di.xml can be used to determine what configuration should be imported to specified service

Proposed service & message structure:
```java
service Catalog { # will be similar for any service 
  rpc importConfig (ImportConfigRequest) returns (ImportConfigResponse)  {}
}

message ImportConfigRequest {
  repeated Config config = 1;
}

message ImportConfigResponse {
  bool status = 1;
  string message = 2;
}

message Config {
  string name = 1;
  string value = 2;
}

```

### PR DESCRIPTION
**Pull request by owlsmage**
Сonfiguration import message broker

### Description (*)
Module with queue consumer that provides possibility for services to subscribe for configuration changes and receive changed configuration

### Related Pull Requests
https://github.com/magento/commerce-data-export/pull/15
https://github.com/magento-commerce/magento2-infrastructure/pull/1395

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/catalog-storefront/issues/427

### Code Review Checklist (*)

See detailed [checklist](https://github.com/magento/storefront-message-broker/blob/develop/dev/docs/projectAgreements/Code-Review-checklist.md)

- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
  - expected results in test verified with data from fixture
- [ ] no backward incompatible changes
- [ ] Class usage: magento/storefront-message-broker repo don't use directly classes from magento/saas-export repo and vise-verse
  - Check composer.json dependencies
- [ ] Legacy code is deleted

### CODE REVIEW

**Comment by mslabko** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@magento run all tests

File: app/code/Magento/ConfigurationImportMessageBroker/etc/module.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+<!--
+/**
+* Copyright © Magento, Inc. All rights reserved.
+* See COPYING.txt for license details.
+*/
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
+    <module name="Magento_ConfigurationImportMessageBroker">
+        <sequence>
+            <module name="Magento_MessageQueue" />
```
Comment (mslabko) [inline app/code/Magento/ConfigurationImportMessageBroker/etc/module.xml]: we can't depend on Magento module inside Message Broker

File: app/code/Magento/ConfigurationImportMessageBroker/Model/Configuration/Import/Mapping/SchemaLocator.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+class SchemaLocator implements \Magento\Framework\Config\SchemaLocatorInterface
+{
+    const CONFIG_FILE_SCHEMA = 'configuration_import_mapping.xsd';
+
+    /**
+     * Path to corresponding XSD file with validation rules for merged config
+     *
+     * @var string
+     */
+    protected $schema;
```
Comment (mslabko) [inline app/code/Magento/ConfigurationImportMessageBroker/Model/Configuration/Import/Mapping/SchemaLocator.php]: please, make it private here and in other places where it's possible

**Review by mslabko — CHANGES_REQUESTED** [review]

File: app/code/Magento/ConfigurationImportMessageBroker/etc/module.xml (hunk 1 above)
Comment (lykhachov) [reply app/code/Magento/ConfigurationImportMessageBroker/etc/module.xml]: removed

File: app/code/Magento/ConfigurationImportMessageBroker/Model/Configuration/Import/Mapping/SchemaLocator.php (hunk 2 above)
Comment (lykhachov) [reply app/code/Magento/ConfigurationImportMessageBroker/Model/Configuration/Import/Mapping/SchemaLocator.php]: replaced

**Comment by owlsmage** [comment]
@magento run Static Tests

**Comment by lykhachov** [comment]
@magento run Static Tests

**Comment by lykhachov** [comment]
@magento run Integration Tests

**Comment by lykhachov** [comment]
@magento run all tests

**Comment by mslabko** [comment]
Semantic Version Checker approved  
```
MINOR | /etc/configuration_import_mapping.xsdMagento_ConfigurationImportMessageBroker:0
```

Integration tests is failed

Regarding error 
```composer.json is not equal to etalon-composer.json (dependencyComposer)
Magento/ConfigurationImportMessageBroker/composer.json```
``` 
Please add mentioned module to the file, by analogy with

https://github.com/magento/storefront-search-ce/blob/b383042ccd1deaf56ca4415ec1e4a31c5448afa4/dev/tests/static/testsuite/Magento/Test/Integrity/_files/blacklist/composer_root_modules_storefront_search_ce.txt

**Comment by RuslanKostiv1** [comment]
@owlsmage please add https://github.com/magento-commerce/magento2-infrastructure/pull/1395 in the list of **Related Pull Requests**

**Comment by RuslanKostiv1** [comment]
@magento run Magento Health Index, Static Tests

**Comment by RuslanKostiv1** [comment]
@magento run Magento Health Index, Static Tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@magento run all tests

**Comment by RuslanKostiv1** [comment]
@owlsmage  @lykhachov there's static files failure:
```
Magento\Framework\Exception\LocalizedException: The following patterns didn't return any result:
app/code/Magento/CatalogMessageBroker/Model/ServiceConnector/Configuration/GrpcConfigurationProvider.php
```
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/storefront-message-broker/pull/4/423a2db3e3ddc9afc9bf482292be029e/Statics/console-error-logs.html

is it outdated information and `dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/blacklist/storefront-message-broker.txt` can be removed?

**Comment by RuslanKostiv1** [comment]
@magento import pr to magento-commerce/storefront-message-broker

**Comment by magento-engcom-team** [comment]
@RuslanKostiv1 the pull request successfully imported.

**Comment by RuslanKostiv1** [comment]
@magento run Static Tests

**Pull request merged**